turbopack/crates/turbo-tasks/src/vc/operation.rs

```diff
@@ -0,0 +1,151 @@
+use std::{fmt::Debug, hash::Hash};
+
+use auto_hash_map::AutoSet;
+use serde::{Deserialize, Serialize};
+
+use crate::{trace::TraceRawVcs, CollectiblesSource, RawVc, TaskInput, Upcast, Vc, VcValueTrait};
+
+#[must_use]
+pub struct VcOperation<T>
+where
+    T: ?Sized + Send,
+{
+    pub(crate) node: Vc<T>,
+}
+
+impl<T: ?Sized + Send> VcOperation<T> {
+    /// Creates a new `VcOperation` from a `Vc`.
+    ///
+    /// The caller must ensure that the `Vc` is not a local task and it points to a a single
+    /// operation.
+    pub fn new(node: Vc<T>) -> Self {
+        // TODO to avoid this runtime check, we should mark functions with `(operation)` and return
+        // a VcOperation directly
+        assert!(
+            !node.is_local(),
+            "VcOperation::new can't be used on local tasks"
+        );
+        Self { node }
+    }
+
+    pub fn connect(self) -> Vc<T> {
+        self.node.node.connect();
+        self.node
+    }
+
+    /// Returns the `RawVc` corresponding to this `Vc`.
+    pub fn into_raw(vc: Self) -> RawVc {
+        vc.node.node
+    }
```

Comment:
Should this also deref to `Vc`? Or do you want the conversions to happen through `.connect()`?

Reply:
The decision was that we should not support implicit derefs, in favor of an explicit conversion via `.connect()`. We need to do a connection here in many cases, as this solves problems for the `State<T>` type. Deref shouldn't have any side-effects (like `connect()` does).